### PR TITLE
Ruby: pattern matching: CFG

### PR DESCRIPTION
**Pull request by aibaars**
This pull request adds the CFG implementation for pattern matching.

To help review here is a [rendered CFG](https://gist.githubusercontent.com/aibaars/3d658fc4c6bdd34dc20a62e4b7065d49/raw/6f2c7ecdca9a1e24a75a2fef98b2b0f0dc1dbb24/case.svg) for the CFG's of [case.rb](https://github.com/github/codeql/pull/7283/files#diff-e0ce958a2b71e898f8051e21222ee5f4e5ffa59b8374552cabb97e45be68cafd)

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll [hunk 1]
```diff
@@ -227,6 +227,8 @@ private predicate inMatchingContext(AstNode n) {
   or
   n instanceof CasePattern
   or
+  n = any(CasePattern c).getDesugared()
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:230]: @hvitved Is this something that should be addressed more generally in the CFG library? I guess we'd want to preserve the completion context when desugaring a node. Should `isValidFor(node) implies isValidFor(node.getDesugared())` hold in general?

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:230]: Yes, that sounds right.

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll]: Does this look right to you? https://github.com/github/codeql/pull/7283/commits/06b3d583e512f11f37f9320517cef9d7f3dbc13f

File: ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll [hunk 2]
```diff
@@ -310,7 +310,7 @@ class MatchingCompletion extends ConditionalCompletion {
     this = TNestedCompletion(_, TMatchingCompletion(value), _)
   }
 
-  override SuccessorType getAMatchingSuccessorType() {
+  override ConditionalSuccessor getAMatchingSuccessorType() {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:313]: Could you change it to `MatchingSuccessor` instead and remove the infix cast below?

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:313]: That does not work for `class NestedMatchingCompletion extends NestedCompletion, MatchingCompletion`, does it?

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:313]: `Completion`s are different from `SuccessorType`s (the latter are never nested).

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:313]: What I meant is that `override BooleanSuccessor getAMatchingSuccessorType()` defined in `NestedMatchingCompletion` would not match the signature if we restricted it to `MatchingSuccessor` in the super class (`MatchingCompletion`)

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/controlflow/internal/Completion.qll:313]: Ah, right, sorry.

**Review by hvitved — APPROVED** [review]

**Pull request merged**